File: cdm-sources/cdm/product/collateral/functions/CheckEligibilityByDetails.java
package cdm.product.collateral.functions;

import cdm.product.collateral.CheckEligibilityResult;
import cdm.product.collateral.CheckEligibilityResult.CheckEligibilityResultBuilder;
import cdm.product.collateral.CollateralCriteria;
import cdm.product.collateral.CollateralTreatment;
import cdm.product.collateral.EligibilityQuery;
import cdm.product.collateral.EligibleCollateralCriteria;
import cdm.product.collateral.EligibleCollateralSpecification;
import com.google.inject.ImplementedBy;
import com.rosetta.model.lib.expression.CardinalityOperator;
import com.rosetta.model.lib.functions.ModelObjectValidator;
import com.rosetta.model.lib.functions.RosettaFunction;
import com.rosetta.model.lib.mapper.MapperC;
import com.rosetta.model.lib.mapper.MapperS;
import java.util.Optional;
import javax.inject.Inject;

import static com.rosetta.model.lib.expression.ExpressionOperators.*;

@ImplementedBy(CheckEligibilityByDetails.CheckEligibilityByDetailsDefault.class)
public abstract class CheckEligibilityByDetails implements RosettaFunction {
	
	@Inject protected ModelObjectValidator objectValidator;
	
	// RosettaFunction dependencies
	//
	@Inject protected CheckCriteria checkCriteria;

	/**
	* @param specification Specifications that determine which collateral meets the eligibility and can be used/posted for delivery. For ICMA usecase - this is the basket(s). For ISDA usecase these are the Eligibility Schedule Lists.
	* @param query The eligibility query defines the criteria to filter the specifications that will be used to find the eligibility result. For ICMA usecase - The asset infomation related to potential collateral available in your inventory you can use for the Repo trade. For ISDA - the questions related to the asset infomation you can post as collateral. For ICMA usecase - The issuer infomation related to the potential collateral available in your inventory you can use for the Repo trade. For ISDA - the questions related to the issuer infomation you can post as collateral.
	* @return eligibilityResult 
	*/
	public CheckEligibilityResult evaluate(EligibleCollateralSpecification specification, EligibilityQuery query) {
		CheckEligibilityResult.CheckEligibilityResultBuilder eligibilityResultBuilder = doEvaluate(specification, query);
		
		final CheckEligibilityResult eligibilityResult;
		if (eligibilityResultBuilder == null) {
			eligibilityResult = null;
		} else {
			eligibilityResult = eligibilityResultBuilder.build();
			objectValidator.validate(CheckEligibilityResult.class, eligibilityResult);
		}
		
		return eligibilityResult;
	}

	protected abstract CheckEligibilityResult.CheckEligibilityResultBuilder doEvaluate(EligibleCollateralSpecification specification, EligibilityQuery query);

	protected abstract MapperC<? extends EligibleCollateralCriteria> filteredCriteria(EligibleCollateralSpecification specification, EligibilityQuery query);

	protected abstract MapperC<? extends EligibleCollateralCriteria> checkedCriteria(EligibleCollateralSpecification specification, EligibilityQuery query);

	public static class CheckEligibilityByDetailsDefault extends CheckEligibilityByDetails {
		@Override
		protected CheckEligibilityResult.CheckEligibilityResultBuilder doEvaluate(EligibleCollateralSpecification specification, EligibilityQuery query) {
			CheckEligibilityResult.CheckEligibilityResultBuilder eligibilityResult = CheckEligibilityResult.builder();
			return assignOutput(eligibilityResult, specification, query);
		}
		
		protected CheckEligibilityResult.CheckEligibilityResultBuilder assignOutput(CheckEligibilityResult.CheckEligibilityResultBuilder eligibilityResult, EligibleCollateralSpecification specification, EligibilityQuery query) {
			eligibilityResult
				.setIsEligible(greaterThan(MapperS.of(checkedCriteria(specification, query).resultCount()), MapperS.of(0), CardinalityOperator.All).get());
			
			eligibilityResult
				.setEligibilityQuery(query);
			
			eligibilityResult
				.setSpecification(specification);
			
			eligibilityResult
				.addMatchingEligibleCriteria(checkedCriteria(specification, query).getMulti());
			
			return Optional.ofNullable(eligibilityResult)
				.map(o -> o.prune())
				.orElse(null);
		}
		
		@Override
		protected MapperC<? extends EligibleCollateralCriteria> filteredCriteria(EligibleCollateralSpecification specification, EligibilityQuery query) {
			return MapperS.of(specification).<EligibleCollateralCriteria>mapC("getCriteria", eligibleCollateralSpecification -> eligibleCollateralSpecification.getCriteria())
				.filterItemNullSafe(item -> item.<CollateralTreatment>map("getTreatment", eligibleCollateralCriteria -> eligibleCollateralCriteria.getTreatment()).<Boolean>map("getIsIncluded", collateralTreatment -> collateralTreatment.getIsIncluded()).get());
		}
		
		@Override
		protected MapperC<? extends EligibleCollateralCriteria> checkedCriteria(EligibleCollateralSpecification specification, EligibilityQuery query) {
			return filteredCriteria(specification, query)
				.filterItemNullSafe(item -> checkCriteria.evaluate(item.<CollateralCriteria>map("getCollateralCriteria", eligibleCollateralCriteria -> eligibleCollateralCriteria.getCollateralCriteria()).get(), query));
		}
	}
}

File: cdm-sources/src/cdm/product/collateral/functions/CheckEligibilityByDetails.py
# pylint: disable=line-too-long, invalid-name, missing-function-docstring, missing-module-docstring, superfluous-parens
# pylint: disable=wrong-import-position, unused-import, unused-wildcard-import, wildcard-import, wrong-import-order, missing-class-docstring
from __future__ import annotations
import sys
import datetime
import inspect
from decimal import Decimal
from rosetta.runtime.utils import *
from rosetta.runtime.func_proxy import replaceable, create_module_attr_guardian
from cdm.product.collateral.EligibilityQuery import EligibilityQuery
from cdm.product.collateral.CheckEligibilityResult import CheckEligibilityResult
from cdm.product.collateral.functions.CheckCriteria import CheckCriteria
from cdm.product.collateral.EligibleCollateralSpecification import EligibleCollateralSpecification

__all__ = ['CheckEligibilityByDetails']


@replaceable
def CheckEligibilityByDetails(specification: EligibleCollateralSpecification, query: EligibilityQuery) -> CheckEligibilityResult:
    """
    This function when presented with a set of Asset and Issuer details to be potentially posted as collateral, will check against a set of given Eligibility Criteria or Basket (EligibleCollateralSpecification) to determine which collateral meets the eligibility and can be used/posted for delivery.
    
    Parameters 
    ----------
    specification : EligibleCollateralSpecification
    Specifications that determine which collateral meets the eligibility and can be used/posted for delivery. For ICMA usecase - this is the basket(s). For ISDA usecase these are the Eligibility Schedule Lists.
    
    query : EligibilityQuery
    The eligibility query defines the criteria to filter the specifications that will be used to find the eligibility result. For ICMA usecase - The asset infomation related to potential collateral available in your inventory you can use for the Repo trade. For ISDA - the questions related to the asset infomation you can post as collateral. For ICMA usecase - The issuer infomation related to the potential collateral available in your inventory you can use for the Repo trade. For ISDA - the questions related to the issuer infomation you can post as collateral.
    
    Returns
    -------
    eligibilityResult : CheckEligibilityResult
    
    """
    self = inspect.currentframe()
    
    
    filteredCriteria = rosetta_filter(rosetta_resolve_attr(rosetta_resolve_attr(self, "specification"), "criteria"), lambda item: rosetta_resolve_attr(rosetta_resolve_attr(self, "treatment"), "isIncluded"))
    checkedCriteria = rosetta_filter(rosetta_resolve_attr(self, "filteredCriteria"), lambda item: CheckCriteria(rosetta_resolve_attr(item, "collateralCriteria"), rosetta_resolve_attr(self, "query")))
    eligibilityResult = _get_rosetta_object('CheckEligibilityResult', 'isEligible', all_elements(rosetta_count(rosetta_resolve_attr(self, "checkedCriteria")), ">", 0))
    eligibilityResult = set_rosetta_attr(rosetta_resolve_attr(self, 'eligibilityResult'), 'eligibilityQuery', rosetta_resolve_attr(self, "query"))
    eligibilityResult = set_rosetta_attr(rosetta_resolve_attr(self, 'eligibilityResult'), 'specification', rosetta_resolve_attr(self, "specification"))
    eligibilityResult.add_rosetta_attr(rosetta_resolve_attr(rosetta_resolve_attr(self, eligibilityResult), 'matchingEligibleCriteria'), rosetta_resolve_attr(self, "checkedCriteria"))
    
    
    return eligibilityResult

sys.modules[__name__].__class__ = create_module_attr_guardian(sys.modules[__name__].__class__)
